src/nrnmpi/nrnmpi_dynam.cpp

```diff
@@ -33,221 +36,222 @@ extern "C" {
-                std::cerr << error << " to RTLD_GLOBAL" << std::endl;
-            }
+    // Found the MPI implementation, `handle` refers to it . Now deduce which
+    // MPI implementation that actually is
+    assert(handle);
+    auto const mpi_implementation = [handle] {
+#ifdef MINGW
+        return "msmpi";
+#else
+        if (dlsym(handle, "ompi_mpi_init")) {
+            // OpenMPI
+            return "ompi";
+        } else if (dlsym(handle, "MPI_SGI_vtune_is_running")) {
+            // Got sgi-mpt. MPI_SGI_init exists in both mpt and hmpt, so we look
+            // for MPI_SGI_vtune_is_running which only exists in the non-hmpt
+            // version.
+            return "mpt";
+        } else {
+            // Assume mpich. Could check for MPID_nem_mpich_init...
+            return "mpich";
         }
-        // Figure out where to find lib[core]nrnmpi{...} libraries.
-        auto const libnrnmpi_prefix = []() -> std::string {
-            if (const char* nrn_home = std::getenv("NRNHOME")) {
-                // TODO: what about windows path separators?
-                return std::string{nrn_home} + "/lib/";
-            } else {
-                // Use the directory libnrniv.so is in
-                return path_prefix_to_libnrniv();
-            }
-        }();
-        // `handle` refers to "libmpi.so", figure out which MPI implementation that
-        // is.
-        auto const mpi_implementation = [handle] {
-            if (dlsym(handle, "ompi_mpi_init")) {
-                // OpenMPI
-                return "ompi";
-            } else if (dlsym(handle, "MPI_SGI_vtune_is_running")) {
-                // Got sgi-mpt. MPI_SGI_init exists in both mpt and hmpt, so we look
-                // for MPI_SGI_vtune_is_running which only exists in the non-hmpt
-                // version.
-                return "mpt";
-            } else {
-                // Assume mpich. Could check for MPID_nem_mpich_init...
-                return "mpich";
-            }
-        }();
-        auto const nrn_mpi_library = libnrnmpi_prefix + "libnrnmpi_" + mpi_implementation + ".so";
-        corenrn_mpi_library = libnrnmpi_prefix + "libcorenrnmpi_" + mpi_implementation + ".so";
-        if (!load_nrnmpi(nrn_mpi_library.c_str(), pmes + strlen(pmes))) {
-            return pmes;
+#endif
+    }();
+
+    // Figure out where to find lib[core]nrnmpi{...} libraries. Older versions
+    // of this code used @loader_path on macOS, which will cause problems in
+    // https://github.com/BlueBrain/CoreNeuron/pull/795 where the code that
+    // calls dlopen(libcorenrnmpi_...) will be in libcorenrnmech.so (in some
+    // model-specific directory) rather than the CoreNEURON installation
+    // directory where libcorenrnmpi_*.so live. Now libcorenrnmech_*.so will be
+    // looked for in the same directory as libnrniv.so, which will be incorrect
+    // if CoreNEURON is built externally with dynamic MPI enabled.
```

Comment:
I think this describes the main change in behaviour with this changeset.

Reply:
Another option would be to put the code responsible for managing dynamic MPI into a new library (`libcoreneuron-mpi.so`) that would be installed alongside the MPI-implementation-specific libraries (`libcorenrnmpi_X.so`), but that might be a bit excessive.